### ISSUE TITLE
Softmax Forward

### ISSUE THREAD

**Issue by ppaleja** [ppaleja/llmsys_f25_hw4#5]
In this part, you will implement a fused kernel of softmax in the attention mechanism.

The softmax function for a vector \\( \\mathbf{x} \\) is defined as:

$\\text{softmax}(\\mathbf{x})\_i = \\frac{e^{x\_i}}{\\sum\_{j} e^{x\_j}}$

where \\( x\_i \\) is the \\( i \\)\-th element of \\( \\mathbf{x} \\).

The kernel also incorporates attention mechanisms, particularly in its handling of attention masks. Attention masks are used to control the model's focus on specific parts of the input.

### Instructions

1.  **Implement the CUDA Kernel**  
    Write the CUDA kernel for softmax in `src/softmax_kernel.cu`:  
    
    ```cpp
      template <typename T, int block_dim, int ele_per_thread>
      __global__ void ker_attn_softmax(T *inp, ...) {
          ...
      }
    
      template <typename T, int block_dim, int ele_per_thread>
      __global__ void ker_attn_softmax_lt32(T *inp, ...) {
          ...
      }
    ```
    
    -   The `ker_attn_softmax_lt32` kernel is already implemented for sequences shorter than 32 and does not require block-level parallelism.
    -   Review the provided implementation of `ker_attn_softmax_lt32` and use it as a reference to implement `ker_attn_softmax`.
2.  **Compile the CUDA File**  
    Compile the CUDA file using the following command:
    
    ```
    nvcc -o minitorch/cuda_kernels/softmax_kernel.so --shared src/softmax_kernel.cu -Xcompiler -fPIC
    ```
    
3.  **Bind the Kernel to miniTorch**  
    In `minitorch/cuda_kernel_ops.py`, bind the kernel by passing the CUDA stream:
    
    ```python
    class CudaKernelOps(TensorOps):
        @staticmethod
        def attn_softmax_fw(inp: Tensor, mask: Tensor):
            ...
    ```
    
    In `minitorch/tensor_functions.py`, define the softmax forward function:
    
    ```python
    class Attn_Softmax(Function):
        @staticmethod
        def forward(ctx: Context, inp: Tensor, mask: Tensor) -> Tensor:
            ...
    ```
    
4.  **Test the Implementation**  
    Run the provided test script and ensure your implementation achieves an average speedup of approximately **6.5×**:
    
    ```
    python kernel_tests/test_softmax_fw.py
    ```
    

---

### Understanding Softmax Forward Kernels

The `ker_attn_softmax_lt32` and `ker_attn_softmax` kernels are optimized for different input sizes:

-   **`ker_attn_softmax_lt32`:**
    
    -   Utilizes warp-level primitives for reduction.
    -   Suitable for smaller input sizes (<32 sequence length).
    -   Efficient parallel reduction without block-wide synchronization.
-   **`ker_attn_softmax`:**
    
    -   Employs block-level reduction techniques.
    -   Suitable for larger input sizes.
    -   Includes two phases of reduction (max and sum) followed by a normalization step with synchronization.

---

#### Algorithmic Steps

The softmax computation in both kernels consists of three main steps:

1.  **Compute Max**
    
    -   Identify the maximum value for normalization to avoid numerical overflow during exponentiation.
2.  **Compute Exponentials and Sum**
    
    -   Calculate the exponentials of normalized values and their sum for final normalization.
3.  **Compute Final Result**
    
    -   Normalize the exponentials by dividing by the sum to obtain softmax probabilities.
    -   Use CUB library's `BlockStore` to minimize memory transactions.

---

### Computing the Maximum Value for Normalization

Both kernels implement the maximum value computation in the same way. Study the implementation in `ker_attn_softmax_lt32`.

#### Steps to Compute the Maximum Value:

1.  **Thread Local Max:**  
    Each thread computes a local maximum:
    
    -   Intermediate values and attention mask adjustments are stored in `val[token_per_reduce][ele_per_thread]`.
    -   The thread-local maximum is recorded in `l_max[token_per_reduce]`, initialized to `REDUCE_FLOAT_INF_NEG`.
2.  **Future Token Masking:**  
    If future tokens are masked, their values are excluded from the maximum computation by setting them to `REDUCE_FLOAT_INF_NEG`.
    
3.  **Attention Mask Adjustment:**  
    Adjust the input value by adding the corresponding attention mask value.
    
4.  **Iterative Updates:**  
    Update the thread-local maximum using `fmaxf`.
    

---

#### Block-Level Reduction for Global Max:

-   **`ker_attn_softmax_lt32`:**
    
    -   Uses a warp-level reduction with a custom `warpReduce` function.
-   **`ker_attn_softmax`:**
    
    -   Uses block-wide reduction with the CUB library's `BlockLoad` and shared memory for synchronization.
